### PR TITLE
add test cases, to do match up test between sentinel and helios.

### PR DESCRIPTION
**Pull request by yinghuang123**
Required items, please complete
   
   Change(s):
   - See guidance below

   Reason for Change(s):
   - See guidance below

   Version Updated:
   - Required only for Detections/Analytic Rule templates
   - See guidance below

   Testing Completed:
   - See guidance below

   Checked that the validations are passing and have addressed any issues that are present:
   - See guidance below

# Guidance <- remove section before submitting
-----------------------------------------------------------------------------------------------------------
## Before submitting this PR please ensure that you have read the following sections and filled out the changes, reason for change and testing complete sections:

Thank you for your contribution to the Microsoft Sentinel Github repo.

> Details of the code changes in your submitted PR.  Providing descriptions for pull requests ensures there is context to changes being made and greatly enhances the code review process.  Providing associated Issues that this resolves also easily connects the reason.
   
   Change(s):
   - Updated syntax for XYZ.yaml

   Reason for Change(s):
   - New schema used for XYZ.yaml
   - Resolves ISSUE #1234

   Version updated:
   - Yes
   - Detections/Analytic Rule templates are required to have the version updated

> The code should have been tested in a Microsoft Sentinel environment that does not have any custom parsers, functions or tables, so that you validate no incorrect syntax and execution functions properly.  If your submission requires a custom parser or function, it must be submitted with the PR.

   Testing Completed:
   - Yes/No/Need Help

_Note: If updating a detection, you must update the version field._

> Before the submission has been made, please look at running the KQL and Yaml Validation Checks locally.
> https://github.com/Azure/Azure-Sentinel#run-kql-validation-locally

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes/No/Need Help
   
   _Note: Let us know if you have tried fixing the validation error and need help._

> References: 
> - [Guidance for Detection checks](https://github.com/Azure/Azure-Sentinel#pull-request-detection-template-structure-validation-check)
> - [General contribution guidance](https://github.com/Azure/Azure-Sentinel/wiki#what-can-you-contribute-and-how-can-you-create-contributions)
> - [PR validation troubleshooting](https://github.com/Azure/Azure-Sentinel#pull-request)

-----------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (eerus) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Do we really need to publish the API key (data that can be used for intrusion)? If we do, then maybe you can limit it to first 4 and last 4 symbols?

**Review by eerus — CHANGES_REQUESTED** [review]
LGTM overall but I left one security comment.

File: Solutions/CohesitySecurity/Playbooks/Cohesity_Close_Helios_Incident/azuredeploy.json.sh
Comment (kishannerella) [inline Solutions/CohesitySecurity/Playbooks/Cohesity_Close_Helios_Incident/azuredeploy.json.sh]: Can we move this files into Tools/ directory instead? This should be generic script which take the name of the playbook as input and deploys the playbook. I don't think we need one such file for every playbook. Also, as much as possible let's try to seperate development folders and test/script folders.

Please add a description on the top of every new file that we add. We should know what the file is supposed to do by looking at the header.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py:58]: Please add a comment to this function and other functions in this file clearly mentioning what they're supposed to do and what the return value is supposed to be. I know that in some cases it's trivial, but let's have some conventions which will help in readability and maintainability.

Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.py:58]: Also, please add a header in the file describing what this file is supposed to do.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: Please remove this function and let the users always use always use the newly created function. We can remove the duplicated code that way.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: Assert that len() is > 0

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: Please rename the variables
startDaysAgos -> startDaysAgo
endDaysAgos -> endDaysAgo

Let the default value for endDaysAgo be '0' as that's what we use.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: Please create a function for 'headers' construction. This is duplicate code between get_alerts_details and get_alerts

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: Consider removing the default value to avoid mistakes in calling this function. We should only use this for a value that we know.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py:62]: Question: When this script is run, are the logs stored somewhere so that it's useful for debugging? Otherwise, it's difficult to debug.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Please move this [down](https://github.com/cohesity/Azure-Sentinel/pull/8/commits/0181cc6743413b2fc1a3ef975b8b38ed4a8f0331#diff-962835dddd48b7def1a7bed8c5b7a8ec6391a111f42c8f1da2cc8f13bbc70cb7R62),  so that  it's clubbed with the testing functions. Infact, please move all the above code to the end of the file so that there's clear seperation between functions and when we're calling.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Could we please validate that none of these values are empty so that we fail fast and the user knows that their environment is not configured?

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py:7]: Remove duplicate?

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: s/subscriptionId/subscription_id

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/helios.py]: We should not be needing to pass around apiKey for every function. Ideally, we should have the functions in this file clubbed into a class which accepts the API key in the constructor and that should be re-used. I'll leave it upto you if you want to do it now or later.

Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: yes, we could have a class which provide get_*() functions to get all the required field.
while this is just some test codes. 
I don't feel very strong to do this abstraction right now.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done, will be visible after next push.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py:7]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: we don't need to validate/test against test codes, this is over test.
the user just need to fill the cohesity.json file, which is also used during cli deployment.           
it is very rarely that the user would leave some attributes empty.

Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: s/apiKey/api_key

Please remove all such references in other files to be consistent.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: done

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: changed to start_days_ago and end_days_ago.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: done

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: done

File: Solutions/CohesitySecurity/Tests/az.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.py:58]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

File: Solutions/CohesitySecurity/Playbooks/Cohesity_Close_Helios_Incident/azuredeploy.json.sh
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Playbooks/Cohesity_Close_Helios_Incident/azuredeploy.json.sh]: just check, till now, for all of the playbooks we have, they just need one common/same parameter, that's PlaybookName.
so, we could use one sh to serve all.

I will just "hide" this file from common/public-visible branch, for my vrun function of vim, will look for a nearby "azuredeploy.json.sh" file to run, when I press F5 key in azuredeploy.json file.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: this is just local printing, which should be fine.
the intruder could still get it from local cohesity.json file.
we need this info to debug, when test case failed.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py]: Please replace with

"This file contains the wrapper functions to invoke Azure cli to interact with Azure Sentinel"

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py]: nit: Please start sentences with a capital letter.

Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.py]: Same for all the lines in the function and other functions.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.py:62]: Could you please answer this question?

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Please move this line towards the end as well.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py]: This function should be moved to az.test.py

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/helios.py]: This is not yet resolved.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py]: This kind of defensive programming is used all across Cohesity. Could you tell me if not having 'apiKey' in the json will exit here immediately? Or will it proceed further with empty value? If it proceeds without any empty value, then I would insist on assertion. If it fails immediately, then I'll leave it up to you whether to add assert on empty or not.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Why is apiKey in camel case but not resource_group or workspace_name? Please try to be consistent.

**Review by kishannerella — COMMENTED** [review]
Mostly LG. Added some minor comments. Will approve once those are resolved.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: apiKey is in data, not variable name. it should be fine.
impossible to be 100% consistent, for apiKey is also used in helios endpoint request header...
    -H 'apiKey: 33e44eac-ce99-46df-7f4e-9ac39446a66e'
could we change that? 

of course I also want to be 100% consistent..
that's why I used apiKey in python code, to be consistent with helios endpoint request header

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: no, it has to be at the beginning.
for I used relative path in code, if move to the end, all those test cases would fail.

Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py]: 1. I'm sorry. I should have been more clear. I didn't mean at the end of the file. You moved all the remaining code to the end of the file based on my comment. Why is this the only line that's not moved?
2. We should not use chdir in code. It's a bad practice. All paths should be based on root dir which should be based on an environment variable, something like COHESITY_SECURITY_ROOT_DIR or an argument to the function.

Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: I respectfully disagree with the idea of defining a root directory for the project. While it may be useful in some cases, it could potentially introduce dependencies and make the code less portable. Additionally, if all paths are based on the root directory, it may limit the script's flexibility and require the script to be run from a specific location. Moreover, adding another environment variable could make the project unnecessarily complicated.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done.

File: Solutions/CohesitySecurity/Tests/helios.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/helios.py]: done

File: Solutions/CohesitySecurity/Tests/az.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.py:62]: this is a one time test code, not daemon or long time running server application.
no need to Permanent the log.
just write to command line output is enough.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: could you see my comments above?

File: Solutions/CohesitySecurity/Tests/az.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.py]: done

File: Solutions/CohesitySecurity/Tests/az.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.py]: I changed it to "This module provides wrapper functions for Azure CLI to interact with Azure Sentinel."
how about this?

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.py]: Please add a description for the file.

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.py]: Is there a reason to name the file starting with a capital letter, not in sync with other files in the same directory?

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.py]: Why is the name 'get_jobId' a mix of snake case and camel case? Same for the other function names as well.

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.py]: Please assert if jobId or clusterIncarnationId or clusterId is empty here after parsing from the data.

I want to understand what happens in the case where the propertyList doesn't have jobId or clusterIncarnationId? Will there be a run time fault? Or will be default initialized to empty string?

File: Solutions/CohesitySecurity/Tests/Alert.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.test.py]: Please add a description to the file.

Also, why does the name of the file start with a capital letter?

File: Solutions/CohesitySecurity/Tests/Alert.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/Alert.test.py]: Same issue as before. Mix of snake case and camel case.

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.py:44]: What happens in case there are 0 incidents? Will this result in a run time fault? Can we add an assert?

File: Solutions/CohesitySecurity/Tests/az.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.py:62]: As long the response is printed, I'm good.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: These lines are big. Could you please split it? Generally, we try to have a 80-char limit for coding. Is there a limit that we already imposed? If not, let's use the 79 char limit from now on. It's used within Cohesity in C++ and Go.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Same comment as before. Line is too long.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py:18]: s/setUp/setup/

File: Solutions/CohesitySecurity/json_parser.sh
Comment (kishannerella) [inline Solutions/CohesitySecurity/json_parser.sh:6]: In the next PR, please update the other variables in cohesity.json and json_parser.sh to follow the snake case like 'api_key'.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Could you please use a more expressive name than vid? I'm not sure what vid means here.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py:119]: Can you please add log messages for each test mentioning that the test is starting and finished successfully so that someone looking at the output can understand the current status of the test?

Also, in general, since this is testing, it's a good idea to print the output of the commands or the results we get from API responses to aid debugging. When you do this, please ensure that the timestamps are also present in the logs. You can do the latter part in a seperate PR if you wish, but please ensure the first paragraph of my comment is implemented in this PR>

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py]: Entire Cohesity code uses this format and yes, things are always run from the root directory. In this case, technically you don't need to do chdir. All you have to do is store the current directory in a variable and use the variable to open files using relative file paths. But it's fine if you want to do it this way.

**Review by kishannerella — CHANGES_REQUESTED** [review]

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py:119]: the first part is done.

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py:18]: after making this suggested change, this function would not be executed......
seems this is syntax restriction...
undo my change.

File: Solutions/CohesitySecurity/Tests/Alert.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/Alert.test.py]: done

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/Alert.py]: done

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/Alert.py]: done

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/Alert.py]: done

File: Solutions/CohesitySecurity/Tests/Alert.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/Alert.py]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py:119]: also applied this change to another test.py

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [inline Solutions/CohesitySecurity/Tests/az.test.py]: Why is the definition test_all_incidents_in_helios and the functions below are not within the scope of the class TestCohesity?

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py]: Missed this?

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py]: Missed this?

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (kishannerella) [reply Solutions/CohesitySecurity/Tests/az.test.py:18]: Ack

**Review by kishannerella — CHANGES_REQUESTED** [review]

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

File: Solutions/CohesitySecurity/Tests/az.test.py
Comment (yinghuang123) [reply Solutions/CohesitySecurity/Tests/az.test.py]: done

**Review by kishannerella — APPROVED** [review]

**Pull request merged**
